### PR TITLE
Drop support for nscd

### PR DESCRIPTION
**Pull request by keszybz**
nscd is known to be racy [1] and it was already deprecated and later dropped in Fedora a while back [1,2]. We don't need to support obsolete stuff in systemd, and the cache in systemd-resolved provides a better solution anyway.

[1] https://fedoraproject.org/wiki/Changes/DeprecateNSCD
[2] https://fedoraproject.org/wiki/Changes/RemoveNSCD

### CODE REVIEW

**Review by yuwata — APPROVED** [review]

**Comment by keszybz** [comment]
testing-farm:fedora-rawhide-x86_64 reports a failure with installation, but I cannot figure out what the failure is. Anyway, this PR is unlikely to cause an issue with installation, so it should be unrelated.

**Comment by bluca** [comment]
This is still shipped in Debian, are there any signs that upstream glibc is deprecating it?

**Comment by keszybz** [comment]
> This is still shipped in Debian, are there any signs that upstream glibc is deprecating it?

My understanding is that upstream glibc hasn't formally deprecated it, but the folks who work on this in Fedora/RHEL are also the upstream maintainers. They don't want to work on this and fix bugs, so effectively it's unmaintained.

**Comment by keszybz** [comment]
The discussion on libc-alpha:
https://sourceware.org/pipermail/libc-alpha/2022-February/136741.html
https://sourceware.org/pipermail/libc-alpha/2022-March/136763.html

> I think it is broadly in agreement with what is discussed here:
> - dlopen-ing arbitrary libraries is very bad
> - static linking should work just as well
> - Keep rump NSCD with caching and other inessential complexity removed

The code in systemd provides support for flushing of caches in nscd, so if nscd loses the caching functionality, we certainly can drop the code in systemd. Nevertheless, the question is whether we should do it now. The code on our side is not intrusive and does not require any maintenance, so maybe we should keep it for now and revisit the subject in a while.

**Comment by yuwata** [comment]
Change itself is good. Maybe, better to announce in NEWS for v256 that we will drop nscd support in v257 ?

**Comment by flokli** [comment]
That might be misleading. You're not dropping support for nscd (that's up to glibc to do), but you're dropping explicit cache flushing, and discouraging people to use glibc-nscd (because it does have a cache).

**Comment by poettering** [comment]
I'd be fine with dropping it now. i'd also be fine going the way @yuwata suggested: announce it in NEWS in v256 and kill it in v257

**Comment by keszybz** [comment]
https://github.com/systemd/systemd/pull/31088 adds the announcement.

**Comment by poettering** [comment]
@keszybz can you rebase? then we can merge this for v257

File: src/core/dynamic-user.c [hunk 1]
```diff
@@ -411,9 +409,9 @@ static int dynamic_user_realize(
                 /* Let's see if a proper, static user or group by this name exists. Try to take the lock on
                  * /etc/passwd, if that fails with EROFS then /etc is read-only. In that case it's fine if we don't
                  * take the lock, given that users can't be added there anyway in this case. */
-                etc_passwd_lock_fd = take_etc_passwd_lock(NULL);
-                if (etc_passwd_lock_fd < 0 && etc_passwd_lock_fd != -EROFS)
-                        return etc_passwd_lock_fd;
+                r = etc_passwd_lock_fd = take_etc_passwd_lock(NULL);
```
Comment (YHNdnzj) [inline src/core/dynamic-user.c:412]: Hmm?

Comment (keszybz) [reply src/core/dynamic-user.c:412]: Do you see something wrong?

**Comment by bluca** [comment]
Don't think I've seen this fail before?

`80/92 systemd:integration-tests / TEST-71-HOSTNAME                                              FAIL             62.49s   exit status 1`

File: src/core/dynamic-user.c (hunk 1 above)
Comment (YHNdnzj) [reply src/core/dynamic-user.c:412]: Ah, this is still assigned to etc_passwd_lock. Previously I overlooked that.

We don't usually use r in such cases, but it's OK.

File: meson_options.txt [hunk 2]
```diff
@@ -55,8 +55,6 @@ option('initrd', type : 'boolean',
        description : 'install services for use when running systemd in initrd')
 option('compat-mutable-uid-boundaries', type : 'boolean', value : false,
        description : 'look at uid boundaries in /etc/login.defs for compatibility')
-option('nscd', type : 'boolean',
```
Comment (YHNdnzj) [inline meson_options.txt:58]: Need to be removed from meson.build too?

Comment (keszybz) [reply meson_options.txt:58]: Yeah, I got distracted halfway.

**Comment by keszybz** [comment]
> Don't think I've seen this fail before?
> 
> `80/92 systemd:integration-tests / TEST-71-HOSTNAME FAIL 62.49s exit status 1`

It just seems to hang when stopping the service. There is no error really, just this:
```[ 7783.235300] H systemd[1]: System is tainted: unmerged-bin
[ 7786.087277] H hostnamectl[160]: Could not set chassis: Invalid chassis 'hoge'
[ 7790.376881] H login[129]: ROOT LOGIN ON pts/0
[ 7790.484983] H systemctl[244]: Job for systemd-hostnamed.service canceled.
[ 7790.517400] H systemd[1]: TEST-71-HOSTNAME.service: Main process exited, code=exited, status=1/FAILURE
[ 7790.517831] H systemd[1]: TEST-71-HOSTNAME.service: Failed with result 'exit-code'.
[ 7790.519486] H systemd[1]: Failed to start TEST-71-HOSTNAME.service - TEST-71-HOSTNAME.
[ 7791.099956] H systemd[1]: Requested transaction contradicts existing jobs: Transaction for systemd-logind.service/start is destruct>
[ 7832.083987] H systemd[1]: systemd-hostnamed.service: State 'stop-sigterm' timed out. Aborting.
[ 7832.084156] H systemd[1]: systemd-hostnamed.service: Killing process 161 (systemd-hostnam) with signal SIGABRT.
[ 7832.084384] H systemd[1]: systemd-hostnamed.service: Killing process 245 (systemd-hostnam) with signal SIGABRT.
[ 7832.107183] H systemd[1]: systemd-hostnamed.service: Main process exited, code=dumped, status=6/ABRT
[ 7832.107542] H systemd[1]: systemd-hostnamed.service: Failed with result 'timeout'.
[ 7832.784742] H systemd[1]: Shutting down.
```

**Review by bluca — APPROVED** [review]

**Pull request merged**